csrc/scheduler/normalization_inner_outer.cpp

```diff
@@ -182,11 +182,239 @@ bool InnerOuterPersistentKernelScheduler::canScheduleCompileTime(
+  int64_t n_elements = tv_buffer_size / data_type_size;
+  int64_t after_vect = n_elements / vectorize_factor;
+  int64_t max_smem = 0;
+  for (int64_t threads = threads_per_block_min;
+       threads <= threads_per_block_max;
+       threads += threads_per_block_step) {
+    int64_t n_batch = ceilDiv(after_vect, threads);
+    max_smem = std::max(
+        max_smem, n_batch * vectorize_factor * threads * data_type_size);
+  }
+  return max_smem;
+}
+
+// Return the broadcast tvs that are broadcast to the iteration dimensions of
+// the inner reduction tv. These tvs are reused in the loop over the iteration
+// dimension. This resue reduced the number loads from gmem and this tensor
+// is likely the first candidate to be moved to shared memory when the register
+// space runs low.
+std::vector<TensorView*> getOuterBroadcastTvs(
+    Fusion* fusion,
+    const std::vector<TensorView*>& reduction_tvs) {
+  // set reference broadcast mask using the first inner reduction tv
+  std::vector<bool> ref_broadcast_mask;
+  for (auto tv : reduction_tvs) {
+    if (scheduler_utils::isFastestDimReduction(tv)) {
+      const auto& root = tv->getMaybeAllocationDomain();
+      ref_broadcast_mask.reserve(root.size());
+      for (const auto i : c10::irange(root.size())) {
+        ref_broadcast_mask.push_back(!root.at(i)->isReduction());
+      }
+      break;
+    }
+  }
+  NVF_ERROR(!ref_broadcast_mask.empty(), "ref_broadcast_mask is empty!");
+
+  // find the brodcast tensor whose broadcast makse is same to the reference
+  std::vector<TensorView*> outer_broadcast_tvs;
+  for (auto tv : ir_utils::allTvs(fusion)) {
+    if (std::any_of(
+            tv->getLeafDomain().begin(),
+            tv->getLeafDomain().end(),
+            [](IterDomain* id) { return id->isBroadcast(); })) {
+      if (auto bcast = dynamic_cast<BroadcastOp*>(tv->definition())) {
+        if (bcast->getBroadcastDimFlags() == ref_broadcast_mask) {
+          outer_broadcast_tvs.emplace_back(tv);
+        }
+      }
+    }
+  }
+  return outer_broadcast_tvs;
+}
+
+// Size of buffers storing intermediate outer reduction results
+int64_t partialOuterReductionBufferSize(
+    const std::vector<TensorView*>& reduction_tvs,
+    SchedulerRuntimeInfo& runtime_info) {
+  int64_t partial_reduction_buffer_size = 0;
+  for (auto buffer : reduction_tvs) {
+    if (scheduler_utils::isFastestDimReduction(buffer)) {
+      continue;
+    }
+    int64_t buffer_size = -1;
+    for (auto id : buffer->getMaybeRFactorDomain()) {
+      if (id->isReduction() || id->isBroadcast()) {
+        continue;
+      }
+      auto id_size = runtime_info.expressionEvaluator().evaluate(id->extent());
+      NVF_ERROR(id_size.hasValue(), "Could not infer persistent buffer size.");
+      if (buffer_size == -1) {
+        buffer_size = id_size.as<int64_t>();
+      } else {
+        buffer_size *= id_size.as<int64_t>();
+      }
+    }
+    buffer_size = (buffer_size == -1) ? 0
```

Comment:
qq: what does `buffer_size == 0` mean? A full reduction/broadcast tensor? shouldn't that still give `buffer_size == 1`

Reply:
basically I'm saying, why not start with buffer_size = 1. so we can get rid of the if condition in line 268

Reply:
It was in the original code to calculate buffer size so I just used it without change.
I think the code only accounts for arrays with size larger than 1 as persistent buffer. For example, almost all the broadcast Tvs can't inline with their consumers, but they are not counted as persistent buffer. It avoids unfriendly values, e.g. when we have a persistent buffer of size 1024, then if plus another 1 for broadcast, it becomes 1025. The definition of persistent buffer is on us and it is only used to estimate register/smem usage, so I belive a slight change should be fine.
I'll investigate this part later and check whether my guess is correct or we can also start with `buffer_size = 1` if nothing bad happens.